api/service/pprof/service.go

```diff
@@ -0,0 +1,235 @@
+			if p.WriteTo(f, profile.Debug); err == nil {
+				utils.Logger().Info().Msg(fmt.Sprintf("Saved profile in: %s", f.Name()))
+			}
+		}
+	} else {
+		utils.Logger().Error().Err(err).Msg(fmt.Sprintf("Could not save profile: %s", profile.Name))
+	}
+}
+
+// handleCpuProfile handles the provided CPU profile by stopping or starting it as specified
+func handleCpuProfile(profile Profile, dir string, shouldStop bool) {
+	if shouldStop {
+		pprof.StopCPUProfile()
+	}
+	f, err := newTempFile(dir, profile.Name, ".pb.gz")
+	if err == nil {
+		pprof.StartCPUProfile(f)
+		utils.Logger().Info().Msg(fmt.Sprintf("Saved CPU profile in: %s", f.Name()))
+	} else {
+		utils.Logger().Error().Err(err).Msg("Could not start CPU profile")
+	}
+}
+
+// unpackProfilesIntoMap unpacks the profiles specified in the configuration into a map
+func (config *Config) unpackProfilesIntoMap() (map[string]Profile, error) {
+	result := make(map[string]Profile)
+	if len(config.ProfileNames) > 0 {
+		for index, name := range config.ProfileNames {
+			if _, ok := result[name]; ok {
+				return nil, errors.New(fmt.Sprintf("Pprof profile names contains duplicate: %v", name))
+			} else {
+				profile := &Profile{
+					Name:     name,
+					Interval: 0,
+					Debug:    0,
+				}
+				// Try set interval value
+				if len(config.ProfileIntervals) == len(config.ProfileNames) {
+					profile.Interval = config.ProfileIntervals[index]
+				} else if len(config.ProfileIntervals) > 0 {
+					profile.Interval = config.ProfileIntervals[0]
+				}
+				// Try set debug value
+				if len(config.ProfileDebugValues) == len(config.ProfileNames) {
+					profile.Debug = config.ProfileDebugValues[index]
+				} else if len(config.ProfileDebugValues) > 0 {
+					profile.Debug = config.ProfileDebugValues[0]
+				}
+				result[name] = *profile
+			}
+		}
+	}
+	return result, nil
+}
+
+func getBasePath(origPath string) string {
+	modifiedPath := strings.Replace(origPath, "\\", "/", -1)
+	if modifiedPath == "" || strings.HasSuffix(modifiedPath, string(os.PathSeparator)) {
+		return filepath.Dir(modifiedPath)
+	} else {
+		return filepath.Dir(modifiedPath + string(os.PathSeparator))
+	}
+}
+
+func joinPath(dir, name string) string {
+	return filepath.Join(append(strings.Split(dir, string(os.PathSeparator)), []string{name}...)...)
+}
+
+// newTempFile returns a new output file in dir with the provided prefix and suffix.
+// Similar to https://github.com/google/pprof/blob/a478d1d731e942fbe69291748caafefe749a19ea/internal/driver/tempfile.go#L25
+func newTempFile(dir, name, suffix string) (*os.File, error) {
+	prefix := name + "."
+	for index := 1; index < 10000; index++ {
+		switch f, err := os.OpenFile(joinPath(dir, fmt.Sprintf("%s%03d%s", prefix, index, suffix)), os.O_RDWR|os.O_CREATE|os.O_EXCL, 0666); {
+		case err == nil:
+			return f, nil
+		case !os.IsExist(err):
+			return nil, err
+		}
+	}
+	// Give up
+	return nil, fmt.Errorf("could not create file of the form %s%03d%s", prefix, 1, suffix)
```

Comment:
What about all 10000 indexes are consumed? Node is a long run program and 10000 is very easy to surpass. Have you considered using other indexes like unix time for the specifier?

Reply:
I agree the [implementation taken from Google](https://github.com/google/pprof/blob/a478d1d731e942fbe69291748caafefe749a19ea/internal/driver/tempfile.go#L25) is not optimal -> now using unix timestamp instead :)

https://github.com/harmony-one/harmony/blob/92568eae89d5cf706b51150ff64e828599d7ed10/api/service/pprof/service.go#L219-L224